### PR TITLE
Mixed Case Name Check Review

### PR DESCRIPTION
**Pull request by Bentleysb**
This check flags objects with name tags that improperly use mixed cases.

Proper case use is defined by set standards and configurable exceptions. 

The standards are as follows:

* Words must start with a capital unless:
    * The first letter is preceded by a number
    * All the words in the name are lower case
* All other letters must be lower case unless: 
    * They follow an apostrophe and, they are not the last letter of the word
    * The entire word is uppercase, except the last letter if it follows or is followed by an apostrophe

The standards are broken by the following configurable exceptions (with default values):

* Articles that are capitalised only if they are the first word:
    * a, an, the
* Prepositions that do not need to start with a capital:
    * and, from, to, of, by, upon, on, off, at, as, into, like, near, onto, per, till, up, via, with, for, in
* Name affixes that may be followed by a capital:
    * Mc, Mac, Mck, Mhic, Mic
* Mixed case units of measurement that are valid after a number:
    * kV
    
These configurables allow this check to be adapted to test different languages.  
OSM uses the `name` tag for the name in a locations primary language, and `name:[ISOcode]` for other languages.
This check uses two configurable to control what languages are checked.

The first is a list of ISO codes for countries that should have there `name` tag checked. It has default values of:

* AIA, ATG, AUS, BHS, BRB, BLZ, BMU, BWA, VGB, CMR, CAN, CYM, DMA, FJI, GMB, GHA, GIB, GRD, GUY, IRL, JAM, KEN, LSO, MWI, MLT, MUS, MSR, NAM, NZL, NGA, PNG, SYC, SLE, SGP, SLB, ZAF, SWZ, TZA, TON, TTO, TCA, UGA, GBR, USA, VUT, ZMB, ZWE

The second is a list of `name:[ISOcode]` tags to check. Default values are:

* name:en

A final configurable is a list of characters that names are split by to for words. Its default values are: 

* SPACE, \-, /, &, @, –

**OSM Example:**
1. Way [id:4780932622](https://www.openstreetmap.org/node/4780932622) has the name `NZ Convenience store`. It is flagged because the S in store should be capitalized. 
2. Way [id:463905525](https://www.openstreetmap.org/way/463905525) has the name:en of `ATM drive through`. This is flagged because drive and through should be capitalized. ATM is fine because it is an all upper case word.

### CODE REVIEW

File: config/configuration.json [hunk 1]
```diff
@@ -131,6 +131,28 @@
       "difficulty": "MEDIUM"
     }
   },
+  "MixedCaseNameCheck": {
+    "check_name.countries":["AIA", "ATG", "AUS", "BHS", "BRB", "BLZ", "BMU", "BWA", "VGB",
```
Comment (danielduhh) [inline config/configuration.json:135]: change to "countries" -> see https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/distributed/IntegrityCheckSparkJob.java#L81

File: config/configuration.json [hunk 2]
```diff
@@ -131,6 +131,28 @@
...
+      "CMR", "CAN", "CYM", "DMA", "FJI", "GMB", "GHA", "GIB", "GRD", "GUY", "IRL", "JAM",
+      "KEN", "LSO", "MWI", "MLT", "MUS", "MSR", "NAM", "NZL", "NGA", "PNG", "SYC", "SLE",
+      "SGP", "SLB", "ZAF", "SWZ", "TZA", "TON", "TTO", "TCA", "UGA", "GBR", "USA", "VUT",
+      "ZMB", "ZWE"],
+    "name.language.keys":["name:en"],
```
Comment (danielduhh) [inline config/configuration.json]: We should logically group these keys; for example
```json
"name":{
  "language.keys",
  "affix",
  "prepositions",
  "articles",
  "units"
}, 
"regex.split": ""
```

File: docs/checks/mixedCaseNameCheck.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    * They follow an apostrophe and, they are not the last letter of the word
+    * The entire word is uppercase, except the last letter if it follows or is followed by an apostrophe
+
+The standards are broken by the following configurable exceptions (with default values):
+
+* Articles that are capitalised only if they are the first word:
+    * a
+    * an
+    * the
+* Prepositions that do not need to start with a capital:
```
Comment (danielduhh) [inline docs/checks/mixedCaseNameCheck.md]: Let's replace these long bullet points with comma separated lists.

File: docs/checks/mixedCaseNameCheck.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+* Mixed case units of measurement that are valid after a number:
+    * kV
+    
+These configurables allow this check to be adapted to test different languages.  
+OSM uses the `name` tag for the name in a locations primary language, and `name:[ISOcode]` for other languages.
+This check uses two configurable to control what languages are checked.
+
+The first is a list of ISO codes for countries that should have there `name` tag checked. It has default values of:
+
+* AIA
```
Comment (danielduhh) [inline docs/checks/mixedCaseNameCheck.md]: See above about comma separated lists

File: docs/checks/mixedCaseNameCheck.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
+# Mixed Case Name Check
```
Comment (danielduhh) [inline docs/checks/mixedCaseNameCheck.md:1]: Continuing the discussion around duplicating code in these high level reviews... I agree with Mike in that we should not have snippets of complex functions, but instead point them to where they exist in the checks code. Instead, we can highlight a few interesting, or essential lines of code we deem relevant.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        final List<String> mixedCaseNameTags = new ArrayList<>();
+        final Map<String, String> osmTags = object.getOsmTags();
+
+        // Check ISO against list of countries for testing name tag
+        if (checkNameCountries.contains(object.tag(ISOCountryTag.KEY).toUpperCase())
+                && Validators.hasValuesFor(object, NameTag.class)
+                && isMixedCase(osmTags.get(NameTag.KEY)))
+        {
+            mixedCaseNameTags.add("name");
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: Use `NameTag.key`

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+        }
+
+        // If mix case id detected, flag
+        if (!mixedCaseNameTags.isEmpty())
+        {
+            this.markAsFlagged(object.getOsmIdentifier());
+            final String osmType;
+            // Get OSM type for object
+            if (object instanceof LocationItem)
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: this can be one line: `String osmFeatureType = object instance of LocationItem ? "Node" : "Way"`

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final List<String> mixedCaseNameTags = new ArrayList<>();
+        final Map<String, String> osmTags = object.getOsmTags();
+
+        // Check ISO against list of countries for testing name tag
+        if (checkNameCountries.contains(object.tag(ISOCountryTag.KEY).toUpperCase())
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: Isn't this ISO code checked in `validCheckForObject` ?

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+     * This function will validate if the supplied atlas object is valid for the check.
+     *
+     * @param object
+     *            the atlas object supplied by the Atlas-Checks framework for evaluation
+     * @return {@code true} if this object should be checked
+     */
+    @Override
+    public boolean validCheckForObject(final AtlasObject object)
+    {
+        return !(object instanceof Relation) && !this.isFlagged(object.getOsmIdentifier())
```
Comment (danielduhh) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: After updating the config to use countries, you can replace this with `object.getTag(ISOCountryTag.KEY).isPresent()`

**Review by danielduhh — CHANGES_REQUESTED** [review]
Great stuff B - here are some minor comments to start

File: config/configuration.json (hunk 1 above)
Comment (Bentleysb) [reply config/configuration.json:135]: `"countries"` will automatically filter out objects that do not have an ISO code in its list?

Comment (danielduhh) [reply config/configuration.json:135]: Yeah - this check shouldn't apply to non-English speaking countries

Comment (Bentleysb) [reply config/configuration.json:135]: It is set up so it can support any language if the configurables are set properly. Any language that it is set to support can then be checked in any country by adding the right tag to `name.language.keys`. This list is to tell the check which countries the set languages are native to, so that it will check the `name` tag as well as the `name:[ISOcode]` tags, in `name.language.keys`, for those countries.

File: docs/checks/mixedCaseNameCheck.md (hunk 5 above)
Comment (Bentleysb) [reply docs/checks/mixedCaseNameCheck.md:1]: How does it look now?

File: config/configuration.json (hunk 1 above)
Comment (danielduhh) [reply config/configuration.json:135]: Would that be conflicting w/ prepositions, articles, and name affix's? Let's say we have `name:es, name:en` in `name.language.keys` that used the same list of articles/prepositions as `name:en`.

Comment (Bentleysb) [reply config/configuration.json:135]: It could be conflicting if a preposition, article, affix, or unit exists in both languages but was only a preposition, article, affix, or unit in one of them. I don't think that should occur very often.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: I'd throw comments between each conditional to briefly (in one line) describe what each one is filtering out.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final List<String> mixedCaseNameTags = new ArrayList<>();
+        final Map<String, String> osmTags = object.getOsmTags();
+
+        // Check ISO against list of countries for testing name tag
+        if (checkNameCountries.contains(object.tag(ISOCountryTag.KEY).toUpperCase())
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: I think this has already been verified in the preliminary filtering. I'd recommend trying to filter as much out as possible in the `validCheckForObject` to try and minimize the runtime of this check.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        for (final String key : languageNameTags)
+        {
+            if (osmTags.containsKey(key) && isMixedCase(osmTags.get(key)))
+            {
+                mixedCaseNameTags.add(key);
+            }
+        }
+
+        // If mix case id detected, flag
+        if (!mixedCaseNameTags.isEmpty())
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java:143]: Why do you need to collect all the objects into an array? You should be able to flag each one as it's read in. Using an array to store all the entities is also really memory intensive (we're dealing with what is essentially big data here) so you should avoid storing objects at all costs.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+                    // If the word is not in the list of prepositions, and the
+                    // word is not both in the article list and not the first word: check that
+                    // the first letter is a capital
+                    if (!lowerCasePrepositions.contains(word)
+                            && !(!firstWord && lowerCaseArticles.contains(word)))
+                    {
+                        final Matcher firstLetterMatcher = Pattern.compile("\\p{L}").matcher(word);
+                        // If the first letter is lower case: return true if it is not preceded by a
+                        // number
+                        if (firstLetterMatcher.find()
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: A good rule of thumb is to never return a boolean from a conditional. You can simplify this logic by just returning the result of the conditional itself.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+                                && Character.isLowerCase(firstLetterMatcher.group().charAt(0))
+                                && !(firstLetterMatcher.start() != 0 && Character
+                                        .isDigit(word.charAt(firstLetterMatcher.start() - 1))))
+                        {
+                            return true;
+                        }
+                    }
+                    // If the word is not all upper case: check if all the letters not following
+                    // apostrophes, unless at the end of the word, are lower case
+                    if (Pattern.compile("\\p{Ll}").matcher(word).find()
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: Same here. Just return the result of the conditional.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+        if (Pattern.compile("\\p{Lu}").matcher(value).find())
+        {
+            // Split into words based on configurable characters
+            final String[] wordArray = value.split("[\\Q" + this.splitCharacters + "\\E]");
+            boolean firstWord = true;
+            // Check each word
+            for (final String word : wordArray)
+            {
+                // Check if the word is intentionally mixed case
+                if (!Pattern
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: Something that will make this check more readable would be to move each of the conditionals into their own helper methods to make this method easier to read. 

You could then have methods like `isAllLowerCase()` etc. which are super descriptive.

**Review by savannahostrowski — CHANGES_REQUESTED** [review]
Great start @Bentleysb! 👨‍💻 

A few comments to help satisfy memory constraints and improve readability.

File: config/configuration.json (hunk 1 above)
Comment (Bentleysb) [reply config/configuration.json:135]: @danielduhh, do you think this is an acceptable edge case? I would like to keep the ability for this to be multilingual, but if that is not going to work I need to deal with it before tackling some of the other changes.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 10 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: When I do this, I get conflicting errors from checkstyle and spotless. Checkstyle wants the comments indented and spotless does not. Should I make an issue on osmlab/atlas-checks for this?

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 11 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: This may be irrelevant, depending on what we decide to do about multi-lingual support. However, as it currently stands this is not necessarily verified in `validCheckForObject`. In `validCheckForObject` it is in an or statement that test if there is ISO tag and name tag or a language specific name tag. It could be that an object only has one of the language specific name tags, so the iso and name tags have to be checked for again here.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 10 above)
Comment (savannahostrowski) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: Mmm, that's odd. I've never had that happen before.

Here's an example of it when it passes all the linters -> https://github.com/osmlab/atlas-checks/blob/dev/src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/DuplicateWaysCheck.java

Maybe it's a matter of keeping all the conditionals on their own line? If you've already tried this way of styling it, let me know and we can chat!

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 12 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java:143]: I am not collecting the objects, just the tag keys that are mixed case. I use this list in the instructions to indicate exactly where the problems are. I could create a string that I keep appending to, instead of a list that I join, if that would be more efficient.

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 13 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: I am just going to leave this as is. I like the readability of the logic flow.

File: docs/checks/mixedCaseNameCheck.md (hunk 5 above)
Comment (danielduhh) [reply docs/checks/mixedCaseNameCheck.md:1]: We can do away with all 3 functions code blocks. You did a great job of explaining whats going on in the intro summary. See comment about replacing large bullet point list w/ a comma separated list

File: config/configuration.json (hunk 1 above)
Comment (danielduhh) [reply config/configuration.json:135]: Yeah, do you have a test for that particular edge case? I think multilingual support is great

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 11 above)
Comment (danielduhh) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: If I understand correctly; the OR conditional checks objects that have a `name`, `iso`, and the country exists in list OR has a language name tag (`name:en`)?

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 6 above)
Comment (danielduhh) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: @Bentleysb

**Review by danielduhh — CHANGES_REQUESTED** [review]
Sorry it's taken forever to keep the review going. I added a few comments and tagged you in an older one

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 13 above)
Comment (savannahostrowski) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: I would really still encourage you to  return the conditionals themselves. This will simplify your code.

**Review by savannahostrowski — CHANGES_REQUESTED** [review]
LGTM - still one outstanding comment.

Please re-consider refactoring anywhere where you are return true or false explicitly. Best practice is to just return the conditional itself. Your code will really benefit in terms of readability by doing this.

File: config/configuration.json (hunk 1 above)
Comment (Bentleysb) [reply config/configuration.json:135]: I have unit tests for multilingual support. I do not have one to test for the edge case because it would fail by design.

File: docs/checks/mixedCaseNameCheck.md (hunk 5 above)
Comment (Bentleysb) [reply docs/checks/mixedCaseNameCheck.md:1]: I think you may be looking at an older version of this.  The long bulleted lists are gone and the functions have been separated. Let me know if we can still get rid of any of the code review portions.

Comment (danielduhh) [reply docs/checks/mixedCaseNameCheck.md:1]: Ah yes I was - looks good, however, still to many code blocks. Pick your favorite out of the remaining. You can keep the descriptions of the others if it makes sense. Great tutorial btw

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java (hunk 13 above)
Comment (Bentleysb) [reply src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java]: I have just been trying to do this, without success. I cannot directly return the values, because if they are false the rest of the `wordArray` values will not be checked. I am essentially creating an any match loop. I did try using `stream` and `anyMatch`, but that does not allow me to use the `firstWord` variable that is used to vary what happens the first time through. So without completely redoing the method I am not sure this is possible. However, I was able to condense a number of the conditionals into one.

**Review by danielduhh — APPROVED** [review]

**Comment by danielduhh** [comment]
When you get a chance, add one or more examples to the PR description

**Review by savannahostrowski — APPROVED** [review]
I think this could still be improved by returning the results of the conditional statements...This may or may not require a pretty significant refactor though.

However, I appreciate the conditional condensing/clean up you did! If you're happy with this, then LGTM! 😸

**Comment by danielduhh** [comment]
@savannahostrowski I may have missed that discussion. Is that regarding the validCheckForObject function ?

File: src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+                            && !(firstLetterMatcher.start() != 0 && Character
+                                    .isDigit(word.charAt(firstLetterMatcher.start() - 1))))
+                            // If the word is not all upper case: check if all the letters not
+                            // following
+                            // apostrophes, unless at the end of the word, are lower case
+                            || (Pattern.compile("\\p{Ll}").matcher(word).find()
+                                    && !isMixedCaseApostrophe(word)
+                                    && isProperNonFirstCapital(word)))
+                    {
+                        return true;
```
Comment (savannahostrowski) [inline src/main/java/org/openstreetmap/atlas/checks/validation/tag/MixedCaseNameCheck.java:201]: @danielduhh yeah, I think the discussion happened when you were in Milan. My comment was this method and specifically this part where we are specifically returning  true and false. It's generally better practice to return the result of the conditional statement because we aren't really doing anything else here except returning. Again, this has been cleaned up a lot. However, my inclination is that we could probably further simplify this logic.

**Comment by Bentleysb** [comment]
@danielduhh, there is one example in the description, but I will see if I can find some of the stranger cases to show.
